lib_async/kafka_async.mli

```diff
@@ -0,0 +1,28 @@
+open Async
+
+type producer
+
+type consumer
+
+type 'a response = ('a, Kafka.error * string) result
+
+val produce :
+  producer ->
+  Kafka.topic ->
+  Kafka.partition ->
+  ?key:string ->
+  string ->
+  unit response Deferred.t
+
+val new_producer : (string * string) list -> producer response
+
+val new_consumer : (string * string) list -> consumer response
+
+val new_topic :
+  producer -> string -> (string * string) list -> Kafka.topic response
+
+val consume : consumer -> topic:string -> Kafka.message Pipe.Reader.t response
```

Comment:
Same remark as for the `produce` function. Besides, the topic should be a value not a string. Otherwise, how to pass options to the topic.

Reply:
I used to use a `Kafka.topic` here, but the type is not all that useful, since it contains a `handle` (`rd_kafka_t` which is already in the `consumer`) and a `rd_kafka_topic_t` but the `rd_kafka_subscribe` API never uses `rd_kafka_topic_t`, instead it uses a `rd_kafka_topic_partition_list_t` which is constructed from topic names as a string. No options can be passed to topics, so the only thing it did was to get the name of the topic as string. So there was no need to build a `Kafka.topic` and indeed creating one made it more confusing because one would expect the options to do something, but rdkafka would never see the options.

In the end I removed the `Kafka.topic` type from here since the type system can't certify that the `handle` in the `consumer` matches the `handle` in the `Kafka.topic` and the code would then just ignore one of these which is somewhat unintiutive (e.g. a `Kafka.topic` for a completely different connection would work just as well, since the code would only ever retrieve the name of the connection as string).

Reply:
I don't follow here. Topics have configuration [options](https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md#topic-configuration-properties). Besides, even if just a string is used behind the scene, I think that it's valuable to have a typed topic value.

The second point about two `handle`s is more concerning. Yes, this would be a bad design, possibly leading to code working for strange reasons. Maybe, the OCaml topic might be a wrapper around a (handler, topic) pair?

Reply:
The thing is, the `rd_kafka_subscribe` C API does not take any options since nowhere in that chain it uses a `rd_kafka_topic_t`, so there is no way to pass these options to the C code, so librdkafka would not know about these options.

In the existing code options on a topic are set using `rd_kafka_topic_conf_t` and then a `rd_kafka_topic_t` is created (which saves the options somewhere internally). Consuming is done (among others) via `rd_kafka_consume_start` which takes this `rd_kafka_topic_t` and can apply the options.

Contrast that with `rd_kafka_subscribe`, which takes a handle and a `rd_kafka_topic_partition_list_t`. That list is created from an empty list and topics are added via `rd_kafka_topic_partition_list_add` which is populated with `const char * topic`. It is unaware of any `rd_kafka_topic_t` and therefore does not have a reference to any `rd_kafka_topic_conf_t`. So requiring a `Kafka.handle` is strange since the options defined there would not be passed to rdkafka when subscribing.

I don't know either why exactly `rd_kafka_topic_partition_list_add` does not just take a `rd_kafka_topic_t` or how options for that balanced group are supposed to be set (the only option that can be set is `int32_t partition`). I can reach out to the rdkafka people and ask why the balanced group subscriptions do not accept any options.